java/dataset/src/main/cpp/jni_wrapper.cc

```diff
@@ -363,6 +366,92 @@ std::shared_ptr<arrow::Buffer> LoadArrowBufferFromByteBuffer(JNIEnv* env, jobjec
   return buffer;
 }
 
+arrow::Status ParseFromBufferImpl(const arrow::Buffer& buf, const std::string& full_name,
+                                  google::protobuf::Message* message) {
+  google::protobuf::io::ArrayInputStream buf_stream{buf.data(),
+                                                    static_cast<int>(buf.size())};
+
+  if (message->ParseFromZeroCopyStream(&buf_stream)) {
+    return arrow::Status::OK();
+  }
+  return arrow::Status::IOError("ParseFromZeroCopyStream failed for ", full_name);
+}
+
+template <typename Message>
+arrow::Result<Message> ParseFromBuffer(const arrow::Buffer& buf) {
+  Message message;
+  ARROW_RETURN_NOT_OK(
+      ParseFromBufferImpl(buf, Message::descriptor()->full_name(), &message));
+  return message;
+}
+
+arrow::Status FromProto(const google::protobuf::Struct& struct_config,
+                        std::unordered_map<std::string, std::string>& out) {
```

Comment:
Please be consistent with whether we use pointers or non-const references for out parameters. (The codebase has historically favored pointers due to following an older Google C++ style guide.)